xcm/xcm-builder/src/origin_aliases.rs

```diff
@@ -14,33 +14,71 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
-use frame_support::traits::{ContainsPair, Get};
+use frame_support::traits::{Contains, ContainsPair};
 use sp_std::marker::PhantomData;
 use xcm::latest::prelude::*;
 
-pub struct RemovePrefixAccountId32<Prefix>(PhantomData<Prefix>);
-impl<Prefix: Get<MultiLocation>> ContainsPair<MultiLocation, MultiLocation>
-	for RemovePrefixAccountId32<Prefix>
+pub struct AliasCase<Origin, Target>(PhantomData<(Origin, Target)>);
+impl<Origin, Target> ContainsPair<MultiLocation, MultiLocation> for AliasCase<Origin, Target>
+where
+	Origin: Contains<MultiLocation>,
+	Target: Contains<MultiLocation>,
 {
 	fn contains(origin: &MultiLocation, target: &MultiLocation) -> bool {
-		if let Ok(appended) = (*target).clone().prepended_with(Prefix::get()) {
-			if appended == *origin {
-				match appended.last() {
-					Some(AccountId32 { .. }) => return true,
-					_ => return false,
-				}
-			}
-		}
-		false
+		Origin::contains(origin) && Target::contains(target)
 	}
 }
 
-pub struct AliasCase<T>(PhantomData<T>);
-impl<T: Get<(MultiLocation, MultiLocation)>> ContainsPair<MultiLocation, MultiLocation>
-	for AliasCase<T>
+pub struct AliasForeignAccountId32<Prefix, Target>(PhantomData<(Prefix, Target)>);
+impl<Prefix: Contains<MultiLocation>, Target: Contains<MultiLocation>>
+	ContainsPair<MultiLocation, MultiLocation> for AliasForeignAccountId32<Prefix, Target>
 {
 	fn contains(origin: &MultiLocation, target: &MultiLocation) -> bool {
-		let (o, t) = T::get();
-		&o == origin && &t == target
+		if let (prefix, Some(account_id @ AccountId32 { .. })) = origin.split_last_interior() {
+			return Prefix::contains(&prefix) &&
+				Target::contains(target) &&
+				Some(&account_id) == target.last()
+		}
+		false
+	}
+}
+
+pub struct SiblingPrefix;
+impl Contains<MultiLocation> for SiblingPrefix {
+	fn contains(t: &MultiLocation) -> bool {
+		match *t {
+			MultiLocation { parents: 1, interior: X1(Parachain(_)) } => true,
+			_ => false,
+		}
 	}
 }
+
+pub struct ParentPrefix;
+impl Contains<MultiLocation> for ParentPrefix {
+	fn contains(t: &MultiLocation) -> bool {
+		match *t {
+			MultiLocation { parents: 1, interior: Here } => true,
+			_ => false,
+		}
+	}
+}
+
+pub struct ChildPrefix;
+impl Contains<MultiLocation> for ChildPrefix {
+	fn contains(t: &MultiLocation) -> bool {
+		match *t {
+			MultiLocation { parents: 0, interior: X1(Parachain(_)) } => true,
+			_ => false,
+		}
+	}
+}
+
+pub struct IsNativeAccountId32;
```

Comment:
These structs are not necessary; they can be created using the `match_types!` macro during configuration.